tutorials/scripting/c_sharp/c_sharp_signals.rst

```diff
@@ -0,0 +1,147 @@
+by step tutorial.
+
+If it is still possible to use signals through the ``Connect``/``Disconnect`` API, C# gives us
+a more idiomatic way into the observer pattern.
+
+Signals as C# events
+--------------------
+
+To provide more type-safety, Godot signals are also all available through events. You can handle
+these events, as any other event, with the ``+=`` and ``-=`` operators.
+
+.. code-block:: csharp
+
+    Timer myTimer = GetNode<Timer>("Timer");
+    myTimer.Timeout += () => GD.Print("Timeout!");
+
+In addition, you can always access signal names associated with a node type through its nested
+``SignalName`` class. This is useful when, for example, you want to await on a signal (see :ref:`doc_c_sharp_differences_yield`).
+
+.. code-block:: csharp
+
+    await ToSignal(GetTree(), SceneTree.SignalName.ProcessFrame);
+
+.. note::
+
+    Godot will take care of disconnecting all the signals you connected through events when your
+    nodes are freed. Meaning that: as you don't need to call ``Disconnect`` on all signals you used
+    ``Connect`` on, you don't need to ``-=`` all the signals you used ``+=`` on.
+
+Custom signals as C# events
+---------------------------
+
+To declare a custom event in your C# script, use the ``[Signal]`` attribute on a delegate type.
+Note that the name of this delegate needs to end with ``EventHandler``.
+
+.. code-block:: csharp
+
+    [Signal]
+    public delegate void SayHelloEventHandler();
+
+    [Signal]
+    public delegate void SayHelloWithNameEventHandler(string myString);
+
+Once this is done, Godot will create the appropriate events automatically behind the scenes. You
+can then use said events as you'd do for any other Godot signal. Note that events are named using
+your delegate's name minus the final ``EventHandler`` part.
+
+.. code-block:: csharp
+
+    public override void _Ready()
+    {
+        SayHello += () => GD.Print("Hello!");
+        SayHelloWithName += SayHelloTo;
+    }
+
+    private void SayHelloTo(string name)
+    {
+        GD.Print($"Hello {name}!");
+    }
+
+.. warning::
+
+    If you want to connect to these signals in the editor, you will need to (re)build the project
+    to see them appear.
+
+    You can click the **Build** button in the upper-right corner of the editor to do so.
+
+Signal emission
+---------------
+
+To emit signals, use the ``EmitSignal`` method. Note that, as for signals defined by the engine,
+your custom signal names are listed under the nested ``SignalName`` class.
+
+.. code-block:: csharp
+
+    public void MyMethodEmittingSignals()
+    {
+        EmitSignal(SignalName.SayHello);
+        EmitSignal(SignalName.SayHelloWithName, "World");
+    }
+
+Signals support arguments of:
+
+* All the `built-in types <https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/built-in-types-table>`_,
+* Classes derived from :ref:`Godot.Object <class_Object>`
+* Collections of those types
```

Comment:
Don't we have conversions from a bunch of .Net collections to Godot ones?

Reply:
The `Godot.Collections` types have constructors that would allow you to convert a .NET collection to a Godot one, yes. However, this conversion won't happen automatically so if you define a signal with a `List<int>` parameter that won't work (should result in an error diagnostic from our source generator).

Reply:
Ah yes, conversion will let you *pass in* the list when making the call, but not define the delegate like that. I'm dumb.